File: etapa3-construcao(geracao-de-cod)/stack2/src/src/main/java/com/hotel/service/ReservaService.java
package com.hotel.service;

import com.hotel.dto.ReservaDTO;
import com.hotel.entity.Hospede;
import com.hotel.entity.Quarto;
import com.hotel.entity.Reserva;
import com.hotel.entity.Reserva.StatusReserva;
import com.hotel.exception.BusinessException;
import com.hotel.exception.ResourceNotFoundException;
import com.hotel.repository.HospedeRepository;
import com.hotel.repository.QuartoRepository;
import com.hotel.repository.ReservaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ReservaService {
    
    private final ReservaRepository reservaRepository;
    private final HospedeRepository hospedeRepository;
    private final QuartoRepository quartoRepository;
    private final QuartoService quartoService;
    
    /**
     * Cria uma nova reserva
     */
    @Transactional
    public ReservaDTO criar(ReservaDTO dto) {
        // Valida datas
        validarDatas(dto.getDataCheckIn(), dto.getDataCheckOut());
        
        // Busca hóspede
        Hospede hospede = hospedeRepository.findById(dto.getHospedeId())
            .orElseThrow(() -> new ResourceNotFoundException("Hóspede não encontrado"));
        
        // Busca quarto
        Quarto quarto = quartoRepository.findById(dto.getQuartoId())
            .orElseThrow(() -> new ResourceNotFoundException("Quarto não encontrado"));
        
        // Verifica se o quarto está disponível
        if (!quarto.getDisponivel()) {
            throw new BusinessException("Quarto não está disponível para reserva");
        }
        
        // Verifica se já existe reserva ativa para este quarto
        if (reservaRepository.existsReservaAtivaByQuartoId(quarto.getId())) {
            throw new BusinessException("Já existe uma reserva ativa para este quarto");
        }
        
        // Cria a reserva
        Reserva reserva = new Reserva();
        reserva.setHospede(hospede);
        reserva.setQuarto(quarto);
        reserva.setDataCheckIn(dto.getDataCheckIn());
        reserva.setDataCheckOut(dto.getDataCheckOut());
        reserva.setStatus(StatusReserva.ATIVA);
        
        // Salva a reserva
        Reserva saved = reservaRepository.save(reserva);
        
        // Atualiza a disponibilidade do quarto
        quartoService.atualizarDisponibilidade(quarto.getId(), false);
        
        return convertToDTO(saved);
    }
    
    /**
     * Lista todas as reservas
     */
    @Transactional(readOnly = true)
    public List<ReservaDTO> listarTodas() {
        return reservaRepository.findAll()
            .stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Lista apenas reservas ativas
     */
    @Transactional(readOnly = true)
    public List<ReservaDTO> listarAtivas() {
        return reservaRepository.findByStatus(StatusReserva.ATIVA)
            .stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Busca uma reserva por ID
     */
    @Transactional(readOnly = true)
    public ReservaDTO buscarPorId(UUID id) {
        Reserva reserva = reservaRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Reserva não encontrada"));
        return convertToDTO(reserva);
    }
    
    /**
     * Atualiza uma reserva existente
     */
    @Transactional
    public ReservaDTO atualizar(UUID id, ReservaDTO dto) {
        Reserva reserva = reservaRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Reserva não encontrada"));
        
        // Valida datas
        validarDatas(dto.getDataCheckIn(), dto.getDataCheckOut());
        
        // Atualiza apenas as datas
        reserva.setDataCheckIn(dto.getDataCheckIn());
        reserva.setDataCheckOut(dto.getDataCheckOut());
        
        Reserva updated = reservaRepository.save(reserva);
        return convertToDTO(updated);
    }
    
    /**
     * Cancela uma reserva
     */
    @Transactional
    public void cancelar(UUID id) {
        Reserva reserva = reservaRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Reserva não encontrada"));
        
        if (reserva.getStatus() != StatusReserva.ATIVA) {
            throw new BusinessException("Apenas reservas ativas podem ser canceladas");
        }
        
        // Atualiza o status da reserva
        reserva.setStatus(StatusReserva.CANCELADA);
        reservaRepository.save(reserva);
        
        // Libera o quarto
        quartoService.atualizarDisponibilidade(reserva.getQuarto().getId(), true);
    }
    
    /**
     * Valida se as datas são válidas
     */
    private void validarDatas(LocalDate checkIn, LocalDate checkOut) {
        if (checkIn == null || checkOut == null) {
            throw new BusinessException("Datas de check-in e check-out são obrigatórias");
        }
        
        if (checkOut.isBefore(checkIn) || checkOut.isEqual(checkIn)) {
            throw new BusinessException("Data de check-out deve ser posterior à data de check-in");
        }
        
        if (checkIn.isBefore(LocalDate.now())) {
            throw new BusinessException("Data de check-in não pode ser no passado");
        }
    }
    
    /**
     * Converte Entity para DTO
     */
    private ReservaDTO convertToDTO(Reserva reserva) {
        ReservaDTO dto = new ReservaDTO();
        dto.setId(reserva.getId());
        dto.setHospedeId(reserva.getHospede().getId());
        dto.setQuartoId(reserva.getQuarto().getId());
        dto.setDataCheckIn(reserva.getDataCheckIn());
        dto.setDataCheckOut(reserva.getDataCheckOut());
        dto.setValorTotal(reserva.getValorTotal());
        dto.setStatus(reserva.getStatus());
        dto.setCriadoEm(reserva.getCriadoEm());
        dto.setAtualizadoEm(reserva.getAtualizadoEm());
        
        // Campos adicionais
        dto.setHospedeNome(reserva.getHospede().getNome() + " " + reserva.getHospede().getSobrenome());
        dto.setQuartoNumero(reserva.getQuarto().getNumero());
        
        return dto;
    }
}

File: etapa3-construcao(geracao-de-cod)/stack1/src/server/src/services/ReservaService.ts
import { ReservaRepository } from '../repositories/ReservaRepository';
import { QuartoRepository } from '../repositories/QuartoRepository';
import { HospedeRepository } from '../repositories/HospedeRepository';
import { CreateReservaDTO, Reserva, UpdateReservaDTO, AppError } from '../types';
import { createReservaSchema, updateReservaSchema } from '../validators/schemas';

export class ReservaService {
  private reservaRepository: ReservaRepository;
  private quartoRepository: QuartoRepository;
  private hospedeRepository: HospedeRepository;

  constructor() {
    this.reservaRepository = new ReservaRepository();
    this.quartoRepository = new QuartoRepository();
    this.hospedeRepository = new HospedeRepository();
  }

  // Criar nova reserva
  async create(data: CreateReservaDTO): Promise<Reserva> {
    // Validar dados com Zod
    const validatedData = createReservaSchema.parse(data);

    // Verificar se hóspede existe
    const hospede = await this.hospedeRepository.findById(validatedData.hospedeId);
    if (!hospede) {
      throw new AppError('Hóspede não encontrado', 404);
    }

    // Verificar se quarto existe
    const quarto = await this.quartoRepository.findById(validatedData.quartoId);
    if (!quarto) {
      throw new AppError('Quarto não encontrado', 404);
    }

    // Verificar se quarto está disponível
    if (!quarto.disponivel) {
      throw new AppError('Quarto não está disponível para reserva', 400);
    }

    // Verificar se quarto já tem reserva ativa
    const hasReservaAtiva = await this.reservaRepository.hasReservaAtiva(validatedData.quartoId);
    if (hasReservaAtiva) {
      throw new AppError('Quarto já possui uma reserva ativa', 400);
    }

    // Calcular número de diárias
    const checkIn = new Date(validatedData.dataCheckIn);
    const checkOut = new Date(validatedData.dataCheckOut);

    // Validar datas
    if (checkOut <= checkIn) {
      throw new AppError('Data de check-out deve ser posterior à data de check-in', 400);
    }

    const diarias = Math.ceil((checkOut.getTime() - checkIn.getTime()) / (1000 * 60 * 60 * 24));
    const valorTotal = diarias * quarto.precoDiaria;

    // Criar reserva
    const reserva = await this.reservaRepository.create({
      hospedeId: validatedData.hospedeId,
      quartoId: validatedData.quartoId,
      dataCheckIn: checkIn,
      dataCheckOut: checkOut,
      valorTotal
    });

    // Atualizar disponibilidade do quarto
    await this.quartoRepository.updateDisponibilidade(validatedData.quartoId, false);

    return reserva;
  }

  // Listar todas as reservas
  async listAll(): Promise<Reserva[]> {
    return await this.reservaRepository.findAll();
  }

  // Listar reservas ativas
  async listAtivas(): Promise<Reserva[]> {
    return await this.reservaRepository.findAtivas();
  }

  // Buscar reserva por ID
  async findById(id: string): Promise<Reserva> {
    const reserva = await this.reservaRepository.findById(id);
    
    if (!reserva) {
      throw new AppError('Reserva não encontrada', 404);
    }

    return reserva;
  }

  // Atualizar reserva
  async update(id: string, data: UpdateReservaDTO): Promise<Reserva> {
    // Validar dados com Zod
    const validatedData = updateReservaSchema.parse(data);

    // Verificar se reserva existe
    const reservaExistente = await this.findById(id);

    let valorTotal = reservaExistente.valorTotal;

    // Se alterar as datas, recalcular valor
    if (validatedData.dataCheckIn || validatedData.dataCheckOut) {
      const checkIn = validatedData.dataCheckIn 
        ? new Date(validatedData.dataCheckIn) 
        : reservaExistente.dataCheckIn;
      
      const checkOut = validatedData.dataCheckOut 
        ? new Date(validatedData.dataCheckOut) 
        : reservaExistente.dataCheckOut;

      if (checkOut <= checkIn) {
        throw new AppError('Data de check-out deve ser posterior à data de check-in', 400);
      }

      const quarto = await this.quartoRepository.findById(reservaExistente.quartoId);
      const diarias = Math.ceil((checkOut.getTime() - checkIn.getTime()) / (1000 * 60 * 60 * 24));
      valorTotal = diarias * quarto!.precoDiaria;
    }

    const updateData: any = {
      valorTotal
    };

    if (validatedData.dataCheckIn) {
      updateData.dataCheckIn = new Date(validatedData.dataCheckIn);
    }
    if (validatedData.dataCheckOut) {
      updateData.dataCheckOut = new Date(validatedData.dataCheckOut);
    }
    if (validatedData.status) {
      updateData.status = validatedData.status;

      // Se cancelar ou concluir reserva, liberar quarto
      if (validatedData.status === 'cancelada' || validatedData.status === 'concluida') {
        await this.quartoRepository.updateDisponibilidade(reservaExistente.quartoId, true);
      }
    }

    return await this.reservaRepository.update(id, updateData);
  }

  // Cancelar reserva
  async cancel(id: string): Promise<Reserva> {
    return await this.update(id, { status: 'cancelada' });
  }
}
